packages/core/README.md

```diff
@@ -301,6 +301,296 @@ mcp.prompt("summarize", {
+  InMemorySessionAdapter, 
+  InMemoryClientRequestAdapter 
+} from "mcp-lite";
+
+const transport = new StreamableHttpTransport({
+  sessionAdapter: new InMemorySessionAdapter({
+    maxEventBufferSize: 1024
+  }),
+  clientRequestAdapter: new InMemoryClientRequestAdapter({
+    defaultTimeoutMs: 30000  // 30 second timeout for server-to-client requests
+  })
+});
+
+const httpHandler = transport.bind(mcp);
+```
+
+The `ClientRequestAdapter` manages pending server-to-client requests (such as elicitation), storing them temporarily while waiting for client responses. This enables the server to pause execution, send a request to the client, and resume once the client provides a response.
+
+### Advanced Example: Cloudflare KV Adapter
+
+For distributed deployments where multiple worker instances might handle different parts of the same session, implement a custom `ClientRequestAdapter` using persistent storage and polling:
+
+```typescript
+import type { ClientRequestAdapter } from "mcp-lite";
+
+interface PendingRequest {
+  timestamp: number;
+  timeoutMs: number;
+  status: 'pending' | 'resolved' | 'rejected';
+  response?: unknown;
+  error?: string;
+}
+
+export class CloudflareKVClientRequestAdapter implements ClientRequestAdapter {
+  private localPending = new Map<string, { 
+    resolve: (value: unknown) => void; 
+    reject: (reason?: unknown) => void;
+    pollInterval?: ReturnType<typeof setInterval>;
+  }>();
+
+  constructor(
+    private kv: KVNamespace,
+    private defaultTimeoutMs: number = 30000,
+    private pollIntervalMs: number = 1000
+  ) {}
+
+  createPending(
+    sessionId: string | undefined,
+    requestId: string | number,
+    options?: { timeout_ms?: number }
+  ): { promise: Promise<unknown> } {
+    const key = `${sessionId ?? ""}:${String(requestId)}`;
+    const timeoutMs = options?.timeout_ms ?? this.defaultTimeoutMs;
+
+    let resolve!: (value: unknown) => void;
+    let reject!: (reason?: unknown) => void;
+    
+    const promise = new Promise<unknown>((res, rej) => {
+      resolve = res;
+      reject = rej;
+    });
+
+    // Store pending request metadata in KV
+    const pendingRequest: PendingRequest = {
+      timestamp: Date.now(),
+      timeoutMs,
+      status: 'pending'
+    };
+
+    this.kv.put(`pending:${key}`, JSON.stringify(pendingRequest), {
+      expirationTtl: Math.ceil(timeoutMs / 1000) + 60 // Extra buffer for cleanup
+    });
+
+    // Store local handlers for this instance
+    const localEntry = { resolve, reject };
+    this.localPending.set(key, localEntry);
+
+    // Start polling for response from other instances
+    const pollInterval = setInterval(async () => {
+      try {
+        const stored = await this.kv.get(`pending:${key}`, 'json') as PendingRequest | null;
+        
+        if (!stored) {
+          // Request was cleaned up, likely timed out
+          this.cleanupLocal(key, new Error('Request not found'));
+          return;
+        }
```

Comment:
nit with this issue: KV is eventually consistent, so the first read might not be there (after your `put` above)

in extreme cases it can take up to 60s for the PUT to be present

but yeah i think anything that writes to a distributed store should have some kind of grace period, or not mistake "key not present" to mean things were already cleaned up (instead of not-yet-propagated)

Reply:
resolved